h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1764,61 +1765,58 @@ private void compactMoveChunks(Iterable<Chunk> move) {
-                movedToEOF = movedToEOF || chunk.block >= originalBlokCount;
+                moveChunk(chunk, sourceAreaBegin, originalFileSize);
             }
+            // update the metadata (hopefully within the file)
+            store(sourceAreaBegin, originalFileSize);
+            sync();
 
+            Chunk chunkToMove = lastChunk;
             long biggestFileSize = getFileLengthInUse();
-            Chunk chunkToMove = null;
-            if (movedToEOF) {
-                // update the metadata (store at the end of the file)
-                reuseSpace = false;
-                store(0, biggestFileSize);
-                sync();
 
-                chunkToMove = lastChunk;
-                biggestFileSize = getFileLengthInUse();
-
-                // now re-use the empty space
-                reuseSpace = true;
-                for (Chunk c : move) {
-                    if (c.block >= originalBlokCount) {
-                        moveChunk(c, 0, 0);
-                    }
+            boolean movedToEOF = false;
+            // move all chunks, which previously did not fit before reserved area
+            // now we can re-use previously reserved area [sourceAreaBegin, originalFileSize),
+            // but need to reserve [originalFileSize, biggestFileSize)
+            for (Chunk c : move) {
+                if (c.block >= originalBlokCount) {
+                    moveChunk(c, originalFileSize, biggestFileSize);
+                    assert c.block < originalBlokCount;
+                    movedToEOF = true;
                 }
             }
-
-            // update the metadata (hopefully within the file)
-            store(movedToEOF ? originalFileSize : sourceAreaLow, biggestFileSize);
-            sync();
-
-            // if all chunks where moved straight to their target locations,
-            // but that last chunk with updated metadata did not fit,
-            // then our hope was unfounded and now we need to move it again
-            assert lastChunk != null;
-            if (!movedToEOF && lastChunk.block * BLOCK_SIZE >= biggestFileSize) {
-                chunkToMove = lastChunk;
-            }
-
-            if (chunkToMove != null &&
-                    fileStore.predictAllocation(chunkToMove.len) + chunkToMove.len < chunkToMove.block &&
-                    moveChunk(chunkToMove, 0, 0)) {
+            if ((movedToEOF || chunkToMove.block >= originalBlokCount) &&
+                    moveChunk(chunkToMove, originalFileSize, biggestFileSize)) {
+                // store a new chunk with updated metadata (hopefully within a file)
                 store(originalFileSize, biggestFileSize);
+                sync();
+                // and if last chunk did not fit within a file, since we can use
+                // previously reserved area [originalFileSize, biggestFileSize) now,
+                // try to move it closer to BOF
+                boolean moved = moveChunkInside(chunkToMove, biggestFileSize);
```

Comment:
Here moveChunkInside() is good method that is used to move the last two chunks and ensure no overlap each other.